### PR TITLE
Biometric Authentication

### PR DESCRIPTION
**Pull request by brandonpage**
- New connected app controlled "Biometric Authentication" lock screen that forces user re-authentication or biometric to unlock.  Auth token is not refreshed when the app is locked.
- Rewrite ScreenLockManager and implement BiometricAuthenticationManager as implementations of new AppLockManager class.
- SalesforceSDKManager returns ScreenLockManager and BiometricAuthenticationManager interfaces implementations instead internal classes.
- Biometric enrollment dialog is shown upon initial login.  Can be displayed again by the app.
- Native biometric unlock button shown by default but can be disabled by the app.

~TODO:~
- [x] Handle web scheme to prompt biometric.
- [x] Get accurate session timeout.
- [x] Revoke old sessions if user re-authenticates before session is expired.

<p float="left">
  
   
  
</p>

### CODE REVIEW

File: libs/SalesforceSDK/AndroidManifest.xml [hunk 1]
```diff
@@ -47,7 +47,7 @@
             android:theme="@style/SalesforceSDK"
             android:launchMode="singleInstance"
             android:screenOrientation="portrait"
-            android:windowSoftInputMode="adjustResize"
+            android:windowSoftInputMode="adjustNothing"
```
Comment (brandonpage) [inline libs/SalesforceSDK/AndroidManifest.xml:50]: This prevents the biometric login (or SP login) button from moving above they keyboard and blocking the UI if the user chooses to type into the webview.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/LoginActivity.java [hunk 2]
```diff
@@ -76,21 +95,30 @@
  *
  * The bulk of the work for this is actually managed by OAuthWebviewHelper class.
  */
-public class LoginActivity extends AccountAuthenticatorActivity
-        implements OAuthWebviewHelperEvents {
+public class LoginActivity extends FragmentActivity
```
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/LoginActivity.java:98]: `AccountAuthenticatorActivity` was deprecated and did not provide much functionality.  I have replicated everything it provided with the `accountAuthenticatorResult` and `mAccountAuthenticatorResponse` variables in this class.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/LoginActivity.java [hunk 3]
```diff
@@ -415,4 +473,109 @@ public void onReceive(Context context, Intent intent) {
             }
         }
     }
+
+    private void presentBiometric() {
```
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/LoginActivity.java:477]: The reset of the class is biometric code that is extremely similar to what is in `ScreenLockActivity` so I will consider creating a single biometric utility class they both use.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java [hunk 4]
```diff
@@ -455,7 +457,18 @@ protected class AuthWebViewClient extends WebViewClient {
...
+            // Remove the Biometric Login button from the Connected App allow denny screen.
+            if (url.contains("frontdoor.jsp")) {
+                final RelativeLayout parentView = (RelativeLayout) view.getParent();
+                if (parentView != null) {
+                    final Button button = parentView.findViewById(R.id.sf__bio_login_button);
+                    if (button != null) {
+                        button.setVisibility(View.INVISIBLE);
+                    }
+                }
+            }
```
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java:469]: There is likely a cleaner way to do this, will try to improve it.  Suggestions welcome.

Comment (wmathurin) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java:469]: I'll probably have to do something similar with the "Login with IDP" button.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java [hunk 5]
```diff
@@ -453,13 +462,31 @@ protected class AuthWebViewClient extends WebViewClient {
...
 
         @Override
         public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
             Uri uri = request.getUrl();
+
+            // Login webview embedded button has sent the signal to show the  biometric prompt.
+            if (uri.toString().equals(BIOMETRIC_PROMPT) && activity != null) {
+                ((LoginActivity) activity).presentBiometric();
+                return true;
+            }
```
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java:488]: This is untested, but _looks_ correct to me.

File: libs/SalesforceSDK/res/values/bootconfig.xml [hunk 6]
```diff
@@ -7,4 +7,5 @@
         <item>api</item>
         <item>web</item>
     </string-array>
+    <string name="remoteAccessConsumerSecret"></string>
```
Comment (wmathurin) [inline libs/SalesforceSDK/res/values/bootconfig.xml:10]: Hmm, I thought bundling the consumer secret in a mobile app was a no-no for security.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java (hunk 4 above)
Comment (wmathurin) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java:472]: So if you reload the web view from the menu, will the button reappear?

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/security/BiometricAuthenticationManager.kt [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            val options = SalesforceSDKManager.getInstance().loginOptions.asBundle()
+            val intent = Intent(ctx, SalesforceSDKManager.getInstance().loginActivityClass)
+            intent.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
+            intent.addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT)
+            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+            intent.putExtras(options)
+            ctx.startActivity(intent)
+            EventsObservable.get().notifyEvent(EventsObservable.EventType.AppLocked)
+
+            getAccountPrefs(user).edit().putString("token", user.refreshToken).apply()
```
[the remark this answers is not in the document]
Comment (brandonpage) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/security/BiometricAuthenticationManager.kt:70]: Woah, this is something leftover from testing.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java (hunk 4 above)
Comment (brandonpage) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.java:472]: `onPageFinished` gets called again if you reload from the menu.

**Review by wmathurin — APPROVED** [review]
LGTM

**Pull request merged**